alphagov/pay-connector
# PP-13051 corporate exemption with Worldpay authorisation request

## WHAT YOU DID
Ensure that if corporate exemptions are enabled and a corporate card is used, then the charges tabel of the connector database is updatd to the value of optimised and the authorisation request to Worldpay includs <exemption type="OP" placement="OPTIMISED"/> element inside the order element

## How to test

- How should it be reviewed?

    - Check the code and test manually if possible.

* What feedback do you need

    - I've made the test on Line 223 of src/test/java/uk/gov/pay/connector/gateway/worldpay/WorldpayAuthoriseHandlerTest.java a parameterised test for readability but realise this may be better as two separate tests?

    - I have read the ticket as 3D secure enabled meaning the requires_3ds column of the gateway_accounts table.
    - I suspect there's a better way of structuring my changes in [WorldpayAuthoriseOrderTemplate.xml](https://github.com/alphagov/pay-connector/pull/5535/files#diff-a18066368d3d8d932d82af7fd4458986a47b0d41cf495b7432c8497767664976)
    - I'm not sure if the naming of the tests in [WorldpayPaymentProviderTest.java](https://github.com/alphagov/pay-connector/pull/5535/files#diff-d64f06829c2303c7615e41d361e4db01df11be2bc86524770d875079b99d79fd) could be made more explicit.
    - Added a logger within `WorldpayPaymentProvider.java` for whether a corporate card has been used as per Krishna's suggestion.

* Updates since previous merge
   -  Clean up src/main/resources/templates/worldpay/WorldpayAuthoriseOrderTemplate.xml . Sadly we need to keep <#if corporateExemptionEnabled || exemptionEngineEnabled> on line 83 otherwise the next if statement doesn't kick in.
   -  Added a logger that registers if a corporate card has been used and includes the `external_id`.
   -  `WorldpayOrderBuilder` cannot be tested directly as it’s an interface and technically has no logic to test. I've added two more tests to `WorldpayAuthoriseTestHandlerTest.java` instead to include checks for the `isCorporateExemptionEnabled` boolean value instead.

* Steps to test

    - Checkout this branch as well as pp-13038-toggle-corporate-exemptions-button pay-toolbox branch (this may not be needed if this branch has been merged by the time of writing).
    - In pay-connector you will need to change the test Worldpay URLs and test threeDsFlexDdcUrls in config.yaml to the worldpay test urls. You can find theses in pay-magic-urls in pay-infra.
    - Run connector from local branch:
`pay local launch --cluster endtoend --local connector`
    - Create a user:
`pay local user`
    - Run toolbox locally and go to the gateway account for the service just created.
    - Go to ‘Manage PSP’ and switch to Worldpay.
    - This will create a worldpay credentials record (in gateway_account_credentials) but the status needs to be changed
to ACTIVE so you can use APIs or a SQL update. Don't forget to add an `active_start_date`. Alternatively you can run the following query `update gateway_account_credentials set state = 'ACTIVE', active_start_date = now()  where id = 2;`.
    - Remember to set `requires_3ds` to true in the DB on the gateway account.
    - Now set the `worldpay_3ds_flex_credentials` using the following query insert into `worldpay_3ds_flex_credentials(id, gateway_account_id, issuer, organisational_unit_id, jwt_mac_key, version, exemption_engine, corporate_exemptions) values (1,1, 'issuer', 'organisational_unit_id', 'jwt_mac_key', 1,  false, false); `. The correct values can be found on the Worldpay developer docs.
    - Now if you `select * from worldpay_3ds_flex_credentials` you can toggle the exemption_engine and corporate_exemption properties as necessary.
    - Now set your account credentials using the selfservice UI by going first to settings and then to "Your PSP - Worldpay" menu. You will need to obtain these from someone with access to pay-low-pass.
    - Now create a new payment link in order to make some test transactions.
    - You can check if CORPORATE or OPTIMISED has been set if you `select * from charges`.
    - You can use the mock card numbers in the card numbers tech docs to test corporate/non-corporate cards https://docs.payments.service.gov.uk/testing_govuk_pay/#mock-card-numbers
    - You should also check that each transaction has a `REQUESTED_3DS_EXEMPTION` event that is either set to `OPTIMISED` or `CORPORATE` dependant upon how you set `exemption_engine`, `corporate_exemptions` and whether you use a corporate card or not.
